design/003-service-topology-support.md

```diff
@@ -0,0 +1,118 @@
+  1. `kubernetes.io/hostname`: routes traffic only to endpoints on the same node;
+  2. `topology.kubernetes.io/zone`: routes traffic to endpoints in the same zone;
+  3. `topology.kubernetes.io/region`: routes traffic to endpoints in the same region.
+
+Finally, there is also a "wildcard" label `*` that implies any topology. Depending on the use case, service owners can combine the four labels in their preference-ordered list to have better control of the service routing.
+
+Linkerd makes its [load balancing](https://linkerd.io/2018/11/14/grpc-load-balancing-on-kubernetes-without-tears/) decisions, so it does not support this feature out of the box. The current implementation of service topology is done at the `kube-proxy` level which resolves the `ClusterIP` of the service to an endpoint chosen based on the `topologyKeys` field of a service. By adding support for this feature, we can get access to valuable metadata related to the topology of a cluster and make use of it when making our own load balancing decisions. Some benefits of doing this would come in the form of advanced routing decisions in multicluster situations, control over egress costs between regions or better handling of failures. 
+
+Toplogy metadata that is obtained from supporting this feature allows for weighting endpoints based on locality.
+
+(_Adapted from [#4325](https://github.com/linkerd/linkerd2/issues/4325)_)
+(_N.B: I use locality and topology interchangeably here, when used in the same context to me both mean the same thing, apologies in advance if the language barrier created any confusion._)
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+### Summary
+---
+
+To implement support for this feature, my main assumption is that most of the work will be carried out in the `destination service` on the control plane side. Although the feature does not add a lot of changes in terms of the Kubernetes API, several changes have to be made. The first and most obvious to me is supporting the `endpointSlices` feature that was introduced in Kubernetes `v1.16`, since this will enable endpoints to hold topology metadata.
+
+All meshed pods have a proxy that will first establish all outbound endpoints of a service the pod is targeting. Through this change, we will introduce a new weight to each endpoint based on locality, which can afterwards be used to add subsequent features that build on locality-awareness. I propose that the weights be added to each endpoint on the proxy side. To ensure backwards compatibility, there should be a separate gRPC call that can be made to the destination service to get a service's topology preference, and the host's (node) topology labels. The call would include the FQDN/IP of a service, and a topology identifier (similar to the context namespace used for service profiles). 
```

Comment:
Should this be an extra call or just a field addition to the existing response?

Reply:
You mention that weights should be added on the proxy side, what actually decides the weight itself?

Reply:
I started up the destination service locally and had a play with the test destination-client (whose existence eluded me until today). To address the questions:

> Should this be an extra call or just a field addition to the existing response

After I had a look at how the response looks like in practice, I realised I made some confusions regarding how endpoint updates work. 

I think it would make more sense to have an additional field (or fields) to the existing response. In my mind, when the connection is first made, as part of `Destination.Get` we should include along with the service FQDN the (source) node identifier I previously mentioned.

In the response, we can add an additional `topology` field that can be populated with the service preference and source node topology label. Each address can include as part of its labels its topology.

For example, based on the output _structure_ of the test destination client, something like this could work:
```
Add:
labels: <addressSet-metrics-labels>
svcTopology: <service topology labels>
nodeTopology: <node topology labels>
- 172.17.0.19:8080
   labels: <control_plane_ns, deployment (...), kubernetes.io/hostname, topology.zone>
```

What I am unsure of is the implications on performance.

> You mention that weights should be added on the proxy side, what actually decides the weight itself?

I think according to the inner-workings of the service topology feature, the weight should be dependent on the preference listed in the spec. If the weights are to be calculated based on a cost function, it would make the most sense to me to have them calculated on the proxy side, otherwise, we would have to keep track of each endpoint that is added/removed. For example, if the topology preference in a service would be:
1. host
2. zone

and for that same service we have 2 endpoints on the same host and one endpoint in the same zone, then they would be calculated based on a cost function which would minimise in this case the network hops/distance. The weights would then be relative to:
a) the topology preference, like previously mentioned
b) the number of endpoints in each topology.

if we add another endpoint in the same host, all of the weights would have to be scaled to account for the change, which would probably be a bit harder to do on the destination service.

I'm not sure if I am overcomplicating this. As a result of my overcomplication, I'm having a bit of a difficulty in coming up with a way to assign weights.

If it's just a case of priorities, then the whole tasks becomes a bit simpler and can be done on the destination service side. Using the same example, all endpoints on the same host would have a priority of 1, same zone priority of 2.

As a side note, your question made me dig a bit deeper into how other service meshes handle load balancing based on locality.

* [Locality weighted load balancing in envoy](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/locality_weight)
* [Priority levels based on locality in envoy](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/priority)

What caught my eye was the priority level associated with each locality. I'm not sure if I'm on the right track here but it was related to what I understood the problem to be.